magento/devdocs#: Logging into log file. New topic. (#5416)
magento/devdocs: _includes/config-guide/custom-logger-handler-examples.md

You can use one of the following approaches for logging into custom file:
- setup custom log file in the `di.xml`
- setup custom file in the custom logger handler class 

## Setup custom log file in the `di.xml`

1. Define custom log file in the virtual type

```xml
<virtualType name="Magento\Payment\Model\Method\VirtualDebug" type="Magento\Framework\Logger\Handler\Base">
    <arguments>
        <argument name="fileName" xsi:type="string">/var/log/payment.log</argument>
    </arguments>
</virtualType>
```

1. Specify handler which will be used

```xml

> <virtualType name="Magento\Payment\Model\Method\VirtualLogger" type="Magento\Framework\Logger\Monolog">

Reviewer: I would ask you to reconsider the "VirtualLogger" name since it might be a little bit confusing. Technically, it's a real logger, just pre-generated. 
It's not a big deal, but might be a point of confusion for some readers.

Reviewer: So I propose some class name without "Virtual" prefix

Author: `Magento\Payment\Model\Method\VirtualLogger` is the example from the [Magento codebase](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Payment/etc/di.xml#L81). But, I've **implemented** your suggestions in a reason of better explanations.

Thank you!

    <arguments>
        <argument name="handlers" xsi:type="array">
            <item name="debug" xsi:type="object">Magento\Payment\Model\Method\VirtualDebug</item>
        </argument>
    </arguments>
</virtualType>
```

> 1. Inject class of virtual logger to `$this->logger`

Reviewer: Could we say that we inject the declared virtual type into our object instead of `$this->logger`? We clarify that it's accessible via $this->logger below

Author: **added**

```xml
<type name="Magento\Payment\Model\Method\Logger">
    <arguments>
        <argument name="logger" xsi:type="object">Magento\Payment\Model\Method\VirtualLogger</argument>
    </arguments>
</type>
```

> 1. Virtual class `Magento\Payment\Model\Method\VirtualLogger` will be injected into `$this->logger` and will log exception messages into `/var/log/payment.log` file. 

Reviewer: ```suggestion
1. Virtual class `Magento\Payment\Model\Method\VirtualLogger` will be injected into the class (`Magento\Payment\Model\Method\Logger` in the example above) and will log exception messages into `/var/log/payment.log` file. 
```
Basically, we cannot be 100% sure that it will be `$this->logger`. We will inject it as a constructor argument with the name `$logger` but the class property name can be declared in the constructor like `$this->errorLogger = $logger`

Author: Yes, you are correct here, @rogyar !

Custom handler class will be injected into error handler which we specify in `di.xml`.

I've fixed information about that.

## Setup custom file in the custom logger handler class

1. Create a class which will log the necessary data. 

Create a new file `app/code/Vendor/ModuleName/Logger/Handler/ErrorHandler.php` with the next content

```php
<?php
/**
 * @author Vendor
 * @copyright Copyright (c) 2019 Vendor (https://www.vendor.com/)
 */
namespace Vendor\ModuleName\Logger\Handler;

use Magento\Framework\Logger\Handler\Base as BaseHandler;
use Monolog\Logger as MonologLogger;

/**
 * Class ErrorHandler
 */
class ErrorHandler extends BaseHandler
{
    /**
     * Logging level
     *
     * @var int
     */
    protected $loggerType = MonologLogger::ERROR;

    /**
     * File name
     *
     * @var string
     */
    protected $fileName = '/var/log/my_custom_logger/error.log';
}
```

1. Define it for the specific handler type in the `di.xml`: 

For example, `Vendor\ModuleName\Logger\Handler\ErrorHandler` will be defined for the error handler.

```xml    

> <virtualType name="MyCustomLogger" type="Magento\Framework\Logger\Monolog">

Reviewer: In the previous example we have named our virtual type as `Magento\Payment\Model\Method\VirtualLogger` here we name it as `MyCustomLogger`. For a reader who is not familiar with the virtual types it might seem like something important. But technically, it's just a unique name. I suggest using the same naming convention for all examples.

Author: **fixed**

   ```php
   <?php
   /**
    * @author Vendor
    * @copyright Copyright (c) 2019 Vendor (https://www.vendor.com/)
    */
   namespace Vendor\ModuleName\Logger\Handler;

   use Magento\Framework\Logger\Handler\Base as BaseHandler;
   use Monolog\Logger as MonologLogger;

   /**
    * Class ErrorHandler
    */
   class ErrorHandler extends BaseHandler
   {
       /**
        * Logging level
        *
        * @var int
        */
       protected $loggerType = MonologLogger::ERROR;

       /**
        * File name
        *
        * @var string
        */
       protected $fileName = '/var/log/my_custom_logger/error.log';
   }
   ```

1. Define it for the specific handler type in the `di.xml`: 

   For example, `Vendor\ModuleName\Logger\Handler\ErrorHandler` will be defined for the error handler:

   ```xml
   <virtualType name="MyCustomLogger" type="Magento\Framework\Logger\Monolog">
       <arguments>
           <argument name="handlers" xsi:type="array">
               <item name="error" xsi:type="object">Vendor\ModuleName\Logger\Handler\ErrorHandler</item>
           </argument>
       </arguments>
   </virtualType>
   ```

>    where `MyCustomLogger` is the unique identifier of [virtual type]({{page.baseurl}}/extension-dev-guide/build/di-xml-file.html#virtual-types).

Reviewer 2: `MyCustomLogger` is a unique identifier.

> 1. Specify a class name that you will use in your custom logger handler.

Reviewer 2: Specify the class name of the custom logger handler in the `type` definition. Use  the  virtual type name from the previous step as an argument for this type.

Delete the "For example" sentence.

Author: You do not understand a logic here. `Vendor\ModuleName\Observer\MyObserver` is a class where custom handler will be injected.

   use Exception;
   use Magento\Framework\Event\ObserverInterface;
   use Magento\Framework\Event\Observer;

   /**
    * Class MyObserver
    */
   class MyObserver implements ObserverInterface
   {
       /**
        * @var PsrLoggerInterface
        */
       private $logger;

       /**
        * MyObserver constructor.
        *
        * @param PsrLoggerInterface $logger
        */
       public function __construct(
           PsrLoggerInterface $logger
       ) {
           $this->logger = $logger;
       }

       /**
        * @param Observer $observer
        */
       public function execute(Observer $observer)
       {
           try {
               // some code goes here
           } catch (Exception $e) {
               $this->logger->error($e->getMessage());
           }
       }
   }
   ```

1. The class `Vendor\ModuleName\Logger\Handler\ErrorHandler` will be injected into the `error` handler of the `$logger` property in the `Vendor\ModuleName\Observer\MyObserver`.

   ```xml
   ...
   <argument name="handlers" xsi:type="array">
       <item name="error" xsi:type="object">Vendor\ModuleName\Logger\Handler\ErrorHandler</item>
   </argument>
   ...
   ```

>    It will log exception messages into `/var/log/my_custom_logger/error.log` file.

Reviewer 2: No indentation. Rewrite the sentence to be similar to the one at the end of the previous section.

You can use one of the following approaches for logging into custom file:

> - setup custom log file in the `di.xml`

Author: **done**

> - setup custom file in the custom logger handler class 

Author: **done**

> ## Setup custom log file in the `di.xml`

Author: **done**

1. Define a custom log file in the virtual type:

```xml
<virtualType name="Magento\Payment\Model\Method\MyCustomDebug" type="Magento\Framework\Logger\Handler\Base">
    <arguments>
        <argument name="fileName" xsi:type="string">/var/log/payment.log</argument>
    </arguments>
</virtualType>
```

> where, `Magento\Payment\Model\Method\MyCustomDebug` is the unique [virtual type]({{page.baseurl}}/extension-dev-guide/build/di-xml-file.html#virtual-types) identifier.

Author: **done**

1. Specify the handler which will be used:

```xml
<virtualType name="Magento\Payment\Model\Method\MyCustomLogger" type="Magento\Framework\Logger\Monolog">
    <arguments>
        <argument name="handlers" xsi:type="array">
            <item name="debug" xsi:type="object">Magento\Payment\Model\Method\MyCustomDebug</item>
        </argument>
    </arguments>
</virtualType>
```

> where, `Magento\Payment\Model\Method\MyCustomLogger` is the unique [virtual type]({{page.baseurl}}/extension-dev-guide/build/di-xml-file.html#virtual-types) identifier.

Author: **done**

1. Inject the declared virtual type (`Magento\Payment\Model\Method\MyCustomLogger`) into our object (`Magento\Payment\Model\Method\Logger`):

```xml
<type name="Magento\Payment\Model\Method\Logger">
    <arguments>
        <argument name="logger" xsi:type="object">Magento\Payment\Model\Method\MyCustomLogger</argument>
    </arguments>
</type>
```

> 1. The virtual class `Magento\Payment\Model\Method\MyCustomLogger` will be injected into the `debug` handler of tbe `$logger` property

Author: **done**

 
```xml
...
  <argument name="handlers" xsi:type="array">
      <item name="debug" xsi:type="object">Magento\Payment\Model\Method\MyCustomDebug</item>
  </argument>
...
``` 

> in the `Magento\Payment\Model\Method\Logger` class and will log exception messages into `/var/log/payment.log` file.

Author: **done**

> 

Author: **done**

## Setup custom file in the custom logger handler class

> 1. Create a class which will log the necessary data. 

Author: **done**

>     Create a new file `app/code/Vendor/ModuleName/Logger/Handler/ErrorHandler.php` with the next content

Author: **done**

    ```php
    <?php
    /**
     * @author Vendor
     * @copyright Copyright (c) 2019 Vendor (https://www.vendor.com/)
     */
    namespace Vendor\ModuleName\Logger\Handler;

    use Magento\Framework\Logger\Handler\Base as BaseHandler;
    use Monolog\Logger as MonologLogger;

    /**
     * Class ErrorHandler
     */
    class ErrorHandler extends BaseHandler
    {
        /**
         * Logging level
         *
         * @var int
         */
        protected $loggerType = MonologLogger::ERROR;

        /**
         * File name
         *
         * @var string
         */
        protected $fileName = '/var/log/my_custom_logger/error.log';
    }
    ```

1. Define it for the specific handler type in the `di.xml`: 

    For example, `Vendor\ModuleName\Logger\Handler\ErrorHandler` will be defined for the error handler:

    ```xml
    <virtualType name="MyCustomLogger" type="Magento\Framework\Logger\Monolog">
        <arguments>
            <argument name="handlers" xsi:type="array">
                <item name="error" xsi:type="object">Vendor\ModuleName\Logger\Handler\ErrorHandler</item>
            </argument>
        </arguments>
    </virtualType>
    ```

>     where, `MyCustomLogger` is the unique identifier of [virtual type]({{page.baseurl}}/extension-dev-guide/build/di-xml-file.html#virtual-types).

Reviewer 2: ```suggestion
    where `MyCustomLogger` is the unique identifier of [virtual type]({{page.baseurl}}/extension-dev-guide/build/di-xml-file.html#virtual-types).
```

Author: **done**

1. Specify a class name that you will use in your custom logger handler.

    For example, you will use a custom logger handler in the `Vendor\ModuleName\Observer\MyObserver` class:

    ```xml
    <type name="Vendor\ModuleName\Observer\MyObserver">
        <arguments>
            <argument name="logger" xsi:type="object">MyCustomLogger</argument>
        </arguments>
    </type>
    ```

> 1. Source code of `Vendor\ModuleName\Observer\MyObserver` class:

Reviewer 2: This isn't a step. Remove the number.

Author: **done**

    ```php
    <?php
    /**
     * @author Vendor
     * @copyright Copyright (c) 2019 Vendor (https://www.vendor.com/)
     */
    declare(strict_types=1);

    namespace Vendor\ModuleName\Observer;

    use Psr\Log\LoggerInterface as PsrLoggerInterface;
    use Exception;
    use Magento\Framework\Event\ObserverInterface;
    use Magento\Framework\Event\Observer;

    /**
     * Class MyObserver
     */
    class MyObserver implements ObserverInterface
    {
        /**
         * @var PsrLoggerInterface
         */
        private $logger;

        /**
         * MyObserver constructor.
         *
         * @param PsrLoggerInterface $logger
         */
        public function __construct(
            PsrLoggerInterface $logger
        ) {
            $this->logger = $logger;
        }

        /**
         * @param Observer $observer
         */
        public function execute(Observer $observer)
        {
            try {
                // some code goes here
            } catch (Exception $e) {
                $this->logger->error($e->getMessage());
            }
        }
    }
    ```

> 1. Class `Vendor\ModuleName\Logger\Handler\ErrorHandler` will be injected into the `error` handler of the `$logger` property

Reviewer 2: ```suggestion
1. The class `Vendor\ModuleName\Logger\Handler\ErrorHandler` will be injected into the `error` handler of the `$logger` property in `Vendor\ModuleName\Observer\MyObserver`.
```

Author: **done**

    ```xml
    ...
     <argument name="handlers" xsi:type="array">
        <item name="error" xsi:type="object">Vendor\ModuleName\Logger\Handler\ErrorHandler</item>
     </argument>
    ...
    ```

>     in the `Vendor\ModuleName\Observer\MyObserver` and will log exception messages into `/var/log/my_custom_logger/error.log` file.

Reviewer 2: ```suggestion
    It will log exception messages into `/var/log/my_custom_logger/error.log` file.
```

Author: **done**
